examples/loader.rs

```diff
@@ -0,0 +1,158 @@
+use deno_core::anyhow::Error;
+use deno_core::futures::FutureExt;
+use deno_core::futures::StreamExt;
+use deno_core::resolve_import;
+use deno_core::FsModuleLoader;
+use deno_core::JsRuntime;
+use deno_core::ModuleLoader;
+use deno_core::ModuleSource;
+use deno_core::ModuleSourceFuture;
+use deno_core::ModuleSpecifier;
+use deno_core::ModuleType;
+use deno_core::RuntimeOptions;
+use eszip::format::Header;
+use eszip::format::HeaderFrame;
+use eszip::load_reqwest;
+use futures::future::poll_fn;
+use futures::task::Poll;
+use std::collections::HashMap;
+use std::marker::Unpin;
+use std::pin::Pin;
+use std::rc::Rc;
+use std::sync::Arc;
+use std::sync::Mutex;
+use tokio::io::AsyncReadExt;
+use tokio::io::AsyncSeekExt;
+use tokio_util::codec::Framed;
+
+pub struct StreamingLoader<R: AsyncReadExt + AsyncSeekExt + Unpin> {
+  // TODO(@littledivy): Use `url::Url`
+  headers: Arc<Mutex<HashMap<String, HeaderFrame>>>,
+  buf: Arc<Mutex<R>>,
+  header_size: Arc<Mutex<usize>>,
+}
+
+impl<R: AsyncReadExt + AsyncSeekExt + Unpin + 'static> ModuleLoader
+  for StreamingLoader<R>
+{
+  fn resolve(
+    &self,
+    specifier: &str,
+    referrer: &str,
+    _is_main: bool,
+  ) -> Result<ModuleSpecifier, Error> {
+    Ok(resolve_import(specifier, referrer)?)
+  }
+
+  fn load(
+    &self,
+    module_specifier: &ModuleSpecifier,
+    _maybe_referrer: Option<ModuleSpecifier>,
+    _is_dyn_import: bool,
+  ) -> Pin<Box<ModuleSourceFuture>> {
+    let headers = Arc::clone(&self.headers);
+    let buf = Arc::clone(&self.buf);
+    let header_size = Arc::clone(&self.header_size);
+    let specifier = module_specifier.to_string();
+    async move {
+      // FIXME: Can this be neater?
+      let (frame, header_size) = poll_fn(|ctx| {
+        match (
+          headers.lock().unwrap().get(specifier.as_str()).cloned(),
+          *header_size.lock().unwrap(),
+        ) {
+          (Some(frame), size) if size != 0 => Poll::Ready((frame, size)),
+          (None, _) | (Some(_), 0) | _ => {
+            // FIXME: Don't wake immediately
+            let waker = ctx.waker().clone();
+            waker.wake();
+            Poll::Pending
+          }
+        }
+      })
+      .await;
+
+      println!("{:?}", frame);
+      let (url, start, size) = match frame {
+        HeaderFrame::Module(specifier, source_ptr, ..) => {
+          (specifier, source_ptr.0, source_ptr.1)
+        }
+        // FIXME: Maybe poll here for source here?
+        HeaderFrame::Redirect(..) => unimplemented!(),
+      };
+
+      let mut buf = buf.lock().unwrap();
+      // FIXME: Offset calculation hack.
+      // Maybe offsets should be from the beginning of the file
+      // instead of the data section?
```

Comment:
Can't do that as then we'd have to do a double pass during eszip building: first figure out the length of the data section by building it once with fake offsets, and then do it again with the real offsets.